syne_tune/blackbox_repository/blackbox_surrogate.py

```diff
@@ -135,70 +240,199 @@ def make_model_pipeline(configuration_space, fidelity_space, model):
+        generalized weighted finite differences to account for that.
 
-        self.surrogate_pipeline = self.make_model_pipeline(
-            configuration_space=self.configuration_space,
-            fidelity_space=self.fidelity_space,
-            model=surrogate,
-        )
-        # todo would be nicer to have this in the feature pipeline
-        if max_samples is not None and max_samples < len(X):
-            random_indices = np.random.permutation(len(X))[:max_samples]
-            self.surrogate_pipeline.fit(
-                X=X.loc[random_indices], y=y.loc[random_indices]
-            )
+        :param prediction: Shape (num_fidelities, num_objectives)
+        :return:
+        """
+        num_fidelities = self.num_fidelities
+        if num_fidelities > 1 and self.fit_differences:
+            spacing, is_contiguous = self._fidelity_spacing()
+            if is_contiguous:
+                spacing = 1
+            for objective_pos in self.fit_differences:
+                prediction_new = np.cumsum(prediction[:, objective_pos] * spacing)
+                prediction[:, objective_pos] = prediction_new
+            return prediction
         else:
-            self.surrogate_pipeline.fit(X=X, y=y)
+            return prediction
+
+    def fit_surrogate(self, X: pd.DataFrame, y: pd.DataFrame) -> Blackbox:
+        """
+        Fits a surrogate model to data from a blackbox. Here, the targets `y` can
+        be a matrix with the number of columns equal to the number of fidelity
+        values (the `predict_curves = True` case).
+        """
+        self.surrogate_pipeline = [
+            self.make_model_pipeline(
+                configuration_space=self.configuration_space,
+                fidelity_space=self.fidelity_space,
+                model=self.surrogate,
+                predict_curves=self.predict_curves,
+            )
+            for _ in range(self.num_seeds)
+        ]
+        y = self._transform_to_finite_differences(y)
+        Xs, ys = self._data_for_seeds(X, y)
+        for pipeline, features, targets in zip(self.surrogate_pipeline, Xs, ys):
+            # todo would be nicer to have this in the feature pipeline
+            num_data = len(features)
+            if self.max_fit_samples is not None and self.max_fit_samples < num_data:
+                random_indices = np.random.permutation(num_data)[: self.max_fit_samples]
+                features = features.loc[random_indices]
+                targets = targets.loc[random_indices]
+            pipeline.fit(X=features, y=targets)
         return self
 
     def _objective_function(
         self,
         configuration: dict,
         fidelity: Optional[dict] = None,
         seed: Optional[int] = None,
-    ) -> dict:
+    ) -> ObjectiveFunctionResult:
+        if self.num_seeds == 1:
+            seed = 0
+        elif seed is not None:
+            assert 0 <= seed < self.num_seeds
```

Comment:
inconsistent behavior. if there is only one seed, it doesn't matter what seed I pass. if there are multiple, an exception might be raised

Reply:
OK